File: stack/0402_remove_k_digits/0402_remove_k_digits.py

class Solution(object):
    def removeKdigits(self, num, k):
        if len(num) <= k:
            return "0"
        
        ret = []
        for i in range(len(num)):
            while k > 0 and ret and ret[-1] > num[i]:
                ret.pop()
                k-=1
            ret.append(num[i])
        while k > 0 and ret:
            ret.pop()
            k -= 1
        
        s = "".join(ret).lstrip("0")
        return s if s else "0"


num = "1432219"
k = 3
res = Solution().removeKdigits(num, k)
print(res)

File: stack/0402_remove_k_digits/0402_remove_k_digits.go
package main 

import (
	"fmt"
)

func removeKdigits(num string, k int) string {
	digits := len(num) - k 
	stack := make([]byte, len(num))
	top := 0

	for i := range num {
		for top > 0 && stack[top-1] > num[i] && k > 0 {
			top--
			k--
		}
		stack[top] = num[i]
		top++
	}

	i := 0
	for i < digits && stack[i] == '0' {
		i++
	}
	if i == digits {
		return "0"
	}
	return string(stack[i:digits])
}



func main() {
	num := "1432219"
	k := 3
	res := removeKdigits(num, k)
	fmt.Println(res)
}